engine/equivalence.ml

```diff
@@ -0,0 +1,85 @@
+
+let compare (src: Merge_accessors.constraint_tree) (dst: Merge_accessors.constraint_tree) : bool =
+  let rec init : Merge_accessors.constraint_tree -> specialized_tree = function
+    | (Leaf _ | Failure) as t -> Terminal t
+    | Node (acc, children, fallback) ->
+      let c = children |> List.map (fun (pi, c_tree) -> Symbolic pi, init c_tree) in
+      let fb = fallback |> Option.map (fun (pi, c_tree) -> Symbolic pi, init c_tree) in
+      SNode (acc, c, fb)
+  in
+  let specialize (pi: domain) (pi': domain) : node_constraint =
+    let intersect = Target_sym_engine.Domain.inter in
+    let is_empty (set:domain) =
+      Target_sym_engine.Domain.Set.is_empty set.int && Target_sym_engine.Domain.Set.is_empty set.tag
+    in
+    match intersect pi pi' with
+    | x when x = pi' ->
+      Resolved true
+    | x when is_empty x ->
+      Resolved false
+    | x -> Symbolic x
+  in
+  let rec _compare (src: Merge_accessors.constraint_tree) (dst: specialized_tree) : bool =
+    let rec trim src_acc src_constraints = function
+      | SNode (_, [], None) -> assert false
+      | Terminal c -> Terminal c
+      | SNode (acc, children, fallback) ->
+        let children' =
+          children
+          |> List.filter_map (fun (pi, s_tree) ->
+              match pi with
+              | Resolved false -> None
+              | Resolved true -> Some (pi, trim src_acc src_constraints s_tree)
+              | Symbolic dst_constraint ->
+                if src_acc = acc then
+                  Some (specialize src_constraints dst_constraint, trim src_acc src_constraints s_tree)
+                else
+                  Some (pi, trim src_acc src_constraints s_tree))
+        in
+        let fallback' =
+          match fallback with
+          | None -> None
+          | Some (pi, s_tree) ->
+            match pi with
+            | Resolved false ->
+              None
+            | Resolved true ->
+              Some (pi, trim src_acc src_constraints s_tree)
+            | Symbolic pi ->
+              Some (specialize src_constraints pi, trim src_acc src_constraints s_tree)
+        in
+        SNode (acc, children', fallback')
+    in
+    let rec compare_terminals : constraint_tree * specialized_tree -> bool = function
+      | (Leaf tbox, Terminal (Leaf tbox')) ->
+        tbox = tbox'
+      | (Failure, Terminal Failure) ->
+        true
+      | (Leaf tbox, SNode (_, (Resolved true, s_tree)::[], None)) ->
+        compare_terminals (Leaf tbox, s_tree)
+      | (Leaf tbox, SNode (_, [], Some (Resolved true, fallback))) ->
+        compare_terminals (Leaf tbox, fallback)
+      | _ ->
+        false
+    in
+    match (src, dst) with
+    | ((Leaf _ | Failure), _) -> compare_terminals (src, dst)
+    | Node (acc, children, _fallback), SNode _ ->
+      children
+      |> List.map (fun (pi, child) ->
+          let dst' = trim acc pi dst in
+          _compare child dst')
+      |> List.exists (Bool.equal false)
+      |> not
```

Comment:
This is `List.for_all (fun (pi, child) -> _compare child (trim acc pi dst)) children`.